RippleCarryAdder/ReferenceImplementation.qs

```diff
@@ -288,4 +288,89 @@ namespace Quantum.Kata.RippleCarryAdder {
+
+    // Task 5.1. Adder Modulo 2ᴺ
+    operation AdderModuloNbits_Reference (a : Qubit[], b : Qubit[], sum : Qubit[]) : Unit is Adj {
+        // Modification of Challenge Solution of Task 1.7 (Last Carry bit isn't calculated)
+        // Sum qubits are used to store the carry bits, and the sum is calculated as they get cleaned up
+        let N = Length(a);
+
+        // Calculate carry bits and Store them in Sum
+        LowestBitCarry_Reference(a[0], b[0], sum[0]);
+        for (i in 1 .. N-1) {
+            HighBitCarry_Reference(a[i], b[i], sum[i - 1], sum[i]);
+        }
+        // No need to calculated last (N+1)th carry bit as Addition is Modulo 2ᴺ
+        
+        // Clean sum qubits (Uncompute carries after they aren't needed) and compute sum
+        for (i in N-1 .. -1 .. 1) {
+            Adjoint HighBitCarry_Reference(a[i], b[i], sum[i - 1], sum[i]); // Restore sum[i] to state 0
+            HighBitSum_Reference(a[i], b[i], sum[i - 1], sum[i]); // Compute sum[i]
+        }
+        Adjoint LowestBitCarry_Reference(a[0], b[0], sum[0]); // Restore sum[0] to state 0
+        LowestBitSum_Reference(a[0], b[0], sum[0]); // Compute sum[i]
+    
+    }
+    
+    // Task 5.2. Two's Complement
+    operation TwosComplement_Reference (a : Qubit[]) : Unit is Adj {
+        // transform a into 2ᴺ - 1 - a - One's Complement
+        ApplyToEachA(X, a);
+        
+        // Do (a+1) mod 2ᴺ
+     
+        X(a[0]); // Increment the least significant bit
+     
+        // Since we are incrementing by One, we flip the next bit only if all previous bits are 0.
+        for(i in 1 .. Length(a)-1){
+            (ControlledOnInt(0,X))(a[0..i-1], a[i]); 
+        }
+    }
+    
+    // Task 5.3. Subtractor Modulo 2ᴺ
+    operation SubtractorModuloNbits_Reference (a : Qubit[], b : Qubit[], diff : Qubit[]) : Unit is Adj {
+        // Transform a into its Two's Complement 2ᴺ - a 
+        TwosComplement_Reference(a);
+        
+        // Add 2ᴺ - a and b to get (2ᴺ + b - a) mod 2ᴺ = (b-a) mod 2ᴺ 
+        AdderModuloNbits_Reference(a,b,diff);
+        
+        // Transform 2ᴺ - a back to a by applying Two's Complement again.  
+        TwosComplement_Reference(a);
+    }
+    
+    // Task 5.4. In Place Adder Modulo 2ᴺ
+    operation InPlaceAdderModuloNbits_Reference (a : Qubit[], b : Qubit[]) : Unit is Adj {
+        // Modification of Solution of Task 3.5 (Last Carry bit isn't saved)
+        
+        let N = Length(a);
+        
+        using (tempCarry = Qubit()) {
+            let carries = [tempCarry] + a;
+
+            // Compute carry bits
+            for (i in 0 .. N-1) {
+                Majority_Reference(a[i], b[i], carries[i]);
+            }
+
+            // No need to save last (N+1)th carry bit as Addition is Modulo 2ᴺ
+            
+            // Restore inputs and ancilla, compute sum
+            for (i in N-1 .. -1 .. 0) {
+                UnMajorityAdd_Reference(a[i], b[i], carries[i]);
+            }
```

Comment:
```suggestion
            ApplyToEachA(UnMajorityAdd_Reference, Reversed(Zipped(a, b, carries)));
```

If you apply these changes the declaration of `N` can be removed.

Reply:
I think here you meant Zipped3 not Zipped